Move hoc stats to a cronjob

caleybrock opened the pull request:
@philbogle before I do this for country_count and company_count, is this what you meant? 

This cronjob that runs `analyze_hoc_activity` gets run every 5 min so I'm a bit confused. Did you mean to make a new similar file that runs less often? This still seems like a heavy amount of reads. 
https://github.com/code-dot-org/code-dot-org/blob/staging/cookbooks/cdo-apps/templates/default/crontab.erb#L68

caleybrock commented:
@laurelfan @wjordan What do you think about reducing the frequency that this gets run? https://github.com/code-dot-org/code-dot-org/blob/staging/cookbooks/cdo-apps/templates/default/crontab.erb#L68

@philbogle and I chatted about it definitely being too frequent for this new query, and probably being to frequent for the old ones too

philbogle commented:
I believe the pages that display this information have a cache lifetime of
one hour so recomputing it every 5 minutes won't be even result in
user-visible changes a lot of the time.   It seems like an hourly refresh
of the query should be sufficiently fresh.

*/5 \* \* \* \* BUNDLE_GEMFILE=/home/ubuntu/production/Gemfile bundle exec
/home/ubuntu/production/bin/cronjob
/home/ubuntu/production/bin/analyze_hoc_activity

A few longer term notes:
- It's already incremental by day, if it could run hourly and be
  incremental by hour that would be good.
- It would be good if it could run against the reporting db rather the main
  db.

On Tue, Nov 3, 2015 at 9:57 PM, Caley Brock notifications@github.com
wrote:

> @laurelfan https://github.com/laurelfan @wjordan
> https://github.com/wjordan What do you think about reducing the
> frequency that this gets run?
> https://github.com/code-dot-org/code-dot-org/blob/staging/cookbooks/cdo-apps/templates/default/crontab.erb#L68
> 
> @philbogle https://github.com/philbogle and I chatted about it
> definitely being too frequent for this new query, and probably being to
> frequent for the old ones too
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/5070#issuecomment-153586228
> .

laurelfan commented:
Some of these (there are other things in the cronjob, not just hoc activity) are shown on dashboard pages that aren't cached for 1 hour. But I don't think any of these usages need to be updated any more than hourly.

They should be easy to run against reporting db, I think some are already.

> On Nov 3, 2015, at 22:06, philbogle notifications@github.com wrote:
> 
> I believe the pages that display this information have a cache lifetime of
> one hour so recomputing it every 5 minutes won't be even result in
> user-visible changes a lot of the time. It seems like an hourly refresh
> of the query should be sufficiently fresh.
> 
> */5 \* \* \* \* BUNDLE_GEMFILE=/home/ubuntu/production/Gemfile bundle exec
> /home/ubuntu/production/bin/cronjob
> /home/ubuntu/production/bin/analyze_hoc_activity
> 
> A few longer term notes:
> - It's already incremental by day, if it could run hourly and be
>   incremental by hour that would be good.
> - It would be good if it could run against the reporting db rather the main
>   db.
> 
> On Tue, Nov 3, 2015 at 9:57 PM, Caley Brock notifications@github.com
> wrote:
> 
> > @laurelfan https://github.com/laurelfan @wjordan
> > https://github.com/wjordan What do you think about reducing the
> > frequency that this gets run?
> > https://github.com/code-dot-org/code-dot-org/blob/staging/cookbooks/cdo-apps/templates/default/crontab.erb#L68
> > 
> > @philbogle https://github.com/philbogle and I chatted about it
> > definitely being too frequent for this new query, and probably being to
> > frequent for the old ones too
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/code-dot-org/code-dot-org/pull/5070#issuecomment-153586228
> > .
> > 
> > —
> > Reply to this email directly or view it on GitHub.

caleybrock commented:
I changed the cronjob to run every hour and updated csedweek.org to use the same values.

Based on what happened this morning, if I merge this in, will the values all be 0 for the first hour? or does a cronjob run immediately once it gets changed?

laurelfan commented:
I don't know what happened this morning, do you?

A cronjob does not run immediately once it gets changed, it runs only when
it is scheduled to in the crontab.

On Wed, Nov 4, 2015 at 10:50 AM, Caley Brock notifications@github.com
wrote:

> I changed the cronjob to run every hour and updated csedweek.org to use
> the same values.
> 
> Based on what happened this morning, if I merge this in, will the values
> all be 0 for the first hour? or does a cronjob run immediately once it gets
> changed?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/5070#issuecomment-153828123
> .

wjordan commented:
Reducing the frequency to once per hour sounds good to me. i don't actually know what exactly caused the `lines_of_code` value to get set to 0 this morning, so I can't say for sure that changing the frequency won't cause it to happen, but I would find it unlikely.

caleybrock commented:
It sounded like somehow the cronjob didn't run, or was stopped and the values were set to their defaults here: https://github.com/code-dot-org/code-dot-org/blob/staging/lib/cdo/properties.rb#L52

Either way, if it only gets run on the hour, there will be a period of time where the default values for total_hoc_count will get used. Right now they are 0, but I could make them our count now (around 70,000) so that for that first hour it will have some value.

philbogle commented:
This still doesn't quite make sense to me;  if the job doesn't run, the
properties table will still exist with stale values, so we shouldn't have
to go to the fallback.

That said, rather than containing zero values, the fallback should some
(increasingly stale) snapshot so they at least are plausible and accurate
as of the date of the snapshot;  e.g.

Caley, do you think you might make that change to ensure that we don't
display an obviously broken zero value on key pages if this situation
happens again?

irb(main):001:0> require 'pp'

=> true

irb(main):002:0> pp Properties.send(:fetch_metrics)

{"created_at"=>"2015-11-04T21:11:40+00:00",

 "created_on"=>"2015-11-04",

 "csedweek_organizers"=>38236,

 "csedweek_teachers"=>24025,

 "csedweek_entire_schools"=>12754,

 "csedweek_students"=>4875091,

 "csedweek_countries"=>356,

 "petition_signatures"=>1922485,

 "lines_of_code"=>8412317091}

You could do the same for fetch_hoc_metrics.  That's quite big so maybe you
don't want to include very country and city just the biggest ones.

On Wed, Nov 4, 2015 at 11:11 AM, Caley Brock notifications@github.com
wrote:

> It sounded like somehow the cronjob didn't run, or was stopped and the
> values were set to their defaults here:
> https://github.com/code-dot-org/code-dot-org/blob/staging/lib/cdo/properties.rb#L52
> 
> Either way, if it only gets run on the hour, there will be a period of
> time where the default values for total_hoc_count will get used. Right now
> they are 0, but I could make them our count now (around 70,000) so that for
> that first hour it will have some value.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/code-dot-org/code-dot-org/pull/5070#issuecomment-153834036
> .

caleybrock commented on lib/cdo/properties.rb:
```diff
@@ -41,25 +41,63 @@ def self.get_user_metrics()
+      "Mexico"=>967736,
+      },
+    'total_hoc_count'=>71132,
+    'hoc_country_totals'=>{},
```
@tanyaparker I'm not sure if defaults are needed for this one since we don't use if for non-US? I can add a few, but that's only going to help for some countries. (same logic with companies)

caleybrock commented:
@philbogle I added default values and company/country counts so this should be ready for review.

philbogle commented on bin/analyze_hoc_activity:
```diff
@@ -150,6 +150,23 @@ def main()
     day += 1
   end
 
+  hoc_country_totals = {}
```
Please add a brief header comment to the effect of "Compute the number of hoc events (the grand total and broken down by company and country)"

philbogle commented on bin/analyze_hoc_activity:
```diff
@@ -150,6 +150,23 @@ def main()
+  total_hoc_count = unique_hoc_events.count
+
+  unique_hoc_events.each do |row|
+    data = JSON.parse(row[:data])
```
I suppose any invalid json is our code bug so it's a feature that this blows up if the data is invalid; just confirming.

caleybrock replied:
I suppose. This line was used in hourofcode_helpers previously and seemed to work fine with our current data.

philbogle commented on lib/cdo/properties.rb:
```diff
@@ -41,25 +41,63 @@ def self.get_user_metrics()
-    'csedweek_countries'=>0,
-    'petition_signatures'=>0,
-    'lines_of_code'=>0,
+    'created_at'=>"2015-11-04T21:11:40+00:00",
```
while you're in the neighborhood, can you add spaces around the || to conform to our style guide?

philbogle commented:
LGTM after comments addressed.
